README.md
# active

![img](https://github.com/fosskers/active/workflows/Tests/badge.svg)
[![Go Report Card](https://goreportcard.com/badge/github.com/fosskers/active)](https://goreportcard.com/report/github.com/fosskers/active)

*Keep your Github Action versions up-to-date. 使用の Github Actions を最新に！*

<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
**Table of Contents**

- [English Guide](#english-guide)
    - [Overview](#overview)
    - [Installation](#installation)
    - [Usage](#usage)
        - [Local Repository](#local-repository)
        - [Batch Updates](#batch-updates)
        - [Automatic PRs](#automatic-prs)
    - [Configuration](#configuration)
        - [OAuth](#oauth)
- [日本語](#日本語)
    - [概要](#概要)
    - [インストール](#インストール)
    - [使用法](#使用法)
        - [手元のリポジトリ](#手元のリポジトリ)
        - [一括処理](#一括処理)
        - [自動的 Pull Request](#自動的-pull-request)
    - [設定](#設定)
        - [OAuth認証](#oauth認証)

<!-- markdown-toc end -->

# English Guide
## Overview

If you use Github CI, you'll recognize fields like this in your Workflow Config:

```yaml
- name: Check out my code
  uses: actions/checkout@v2
```

But wait a minute, these Actions [receive
updates](https://github.com/actions/checkout/releases), so our configs can fall
behind!

`active` scans your projects and queries Github for the latest Action releases,
and updates your configs for you:

```
> active --local
Checking the following files:
  --> .: go.yml

Updates available for .: go.yml:
  actions/setup-go 2 --> 2.0.3
  actions/checkout 2 --> 2.1.0
Would you like to apply them? [Y/n] Y
Updated.
```

## Installation

Assuming you have a [Golang environment set up](https://golang.org/doc/install):

```bash
go get github.com/fosskers/active
```

## Usage

### Local Repository

The simplest usage:

```
active --local
```

This will look for workflow files in `./.github/workflows/`.

### Batch Updates

`active` is meant to be configured and ran on multiple projects at once.
Assuming you've configured it (see below), the default usage "just works":

```
> active
Checking the following files:
  --> aura:     ci.yaml
  --> org-mode: ci.yaml
  --> versions: ci.yaml

... etc ...
```

If you trust `active` to do the right thing, you can use `active -y` to
automatically accept all available updates.

### Automatic PRs

With the `--push` flag, `active` will automatically make a commit on a new
branch, push it to Github, and open a PR:

```
> active --push
Checking the following files:
  --> aura:     ci.yaml
  --> org-mode: ci.yaml
  --> versions: ci.yaml

... work ...

Successfully opened a PR for versions! (#35)
Successfully opened a PR for org-mode! (#15)
Successfully opened a PR for aura! (#314)
```

This requires a valid **Personal Access Token** from Github (see below), and
will also create a new Git *remote* called `active` for each project to ensure
that the token can be used properly for pushing.

## Configuration

A config file is not necessary to use `active`, but having one will make your
life easier. By default, `active` looks for its config file at
`$HOME/.config/active.yaml`. Its contents should look like this:

```yaml
projects:
  - /home/you/code/some-project
  - /home/you/code/another-project
  - /home/you/code/third-project

git:
  name:  Your Name      # (Optional) For --push
  email: you@email.com  # (Optional) For --push
  user:  you            # (Optional) For --push
  token: <oauth-token>  # (Optional) For --push, and higher API rate limits in general.
```

`name` and `email` are used for commiting. `user` is used for branch pushing,
and `token` for opening the PR.

If you want to specify an alternate config location, use `--config`.

### OAuth

If you have a Github account, then it's easy to generate a personal access token
for `active`. First visit the [Token
Settings](https://github.com/settings/tokens) on Github. Click **Generate new
token**, and give it `public_repo` permissions:

![](token.png)

# 日本語
## 概要

Github CIの使用の方は下記の設定に見覚えはあるでしょう：

```yaml
- name: Check out my code
  uses: actions/checkout@v2
```

しかしActionsとは [見えない所でアップデートが現れ
る](https://github.com/actions/checkout/releases)物で、こういったアップデートを
見逃してしまう事があります。

`active`はプロジェクトを分析し、使用のActionsを最新のバージョンに更新します：

```
> active --local
Checking the following files:
  --> .: go.yml

Updates available for .: go.yml:
  actions/setup-go 2 --> 2.0.3
  actions/checkout 2 --> 2.1.0
Would you like to apply them? [Y/n] Y
Updated.
```

## インストール

[Golangの開発環境](https://golang.org/doc/install)があれば:

```bash
go get github.com/fosskers/active
```

## 使用法

### 手元のリポジトリ

もっとも簡単な使い方としては：

```
active --local
```

`./.github/workflows/`で見つかる「workflow」ファイルが分析されます。

### 一括処理

本来`active`はファイルで設定され、一括に履行するようにできています。

```
> active
Checking the following files:
  --> aura:     ci.yaml
  --> org-mode: ci.yaml
  --> versions: ci.yaml

... などなど ...
```

設定自体は下記を参考にしてください。

更に`active -y`("yes")で実行すれば、ユーザーの確認を得ずに全ての更新は自動的にさ
れます。

### 自動的 Pull Request

`--push`を加えると`active`は自動的にコミットを作り、ブランチをGithubに送り、新し
いPRが開かれます。

```
> active --push
Checking the following files:
  --> aura:     ci.yaml
  --> org-mode: ci.yaml
  --> versions: ci.yaml

... work ...

Successfully opened a PR for versions! (#35)
Successfully opened a PR for org-mode! (#15)
Successfully opened a PR for aura! (#314)
```

ただしこの場合はGithubからの**Personal Access Token**が必要となります(下記を参考
に)。また、そのTokenが正確に使えるようにプッシュの前、`active`という新しい「Git
remote」が各レポジトリで登録されます。

## 設定

`active`を使うには設定ファイルが特に必要ありませんが、あった方では後が色々と楽になります。
ファイル自体は`$HOME/.config/active.yaml`に置きましょう。内容は以下の様に：

```yaml
projects:
  - /home/daisuke/code/some-project
  - /home/daisuke/code/another-project
  - /home/daisuke/code/third-project

# `--push`を使う時にのみ必須。
git:
  name:  Yamada Daisuke
  email: yamadad@japan.jp
  user:  daisuke
  token: <oauth-token>
```

`--config`にて他の設定ファイルを指定できます。

### OAuth認証

Personal Access Tokenを作るのは簡単です。まずはGithubの[Token
Settings](https://github.com/settings/tokens)のページへ。**Generate new token**
をクリックし、`public_repo`許可のみ与えたら完成です:

![](token.png)


CHANGELOG.md
# `active` Changelog

## 1.0.2 (2020-05-28)

#### Fixed

- Fixed a deadlock and a segfault.

## 1.0.1 (2020-05-25)

#### Changed

- When using `--pull`, the `master` branch is now pulled from the detected
  remote before analysis occurs.
- Performance improvements.

## 1.0.0 (2020-05-20)

This is the initial release of `active`.

#### Added

- Auto-detection of outdated Action versions.
- Config file support. Expected at `$HOME/.config/active.yaml`, but can be
  overridden with `--config`.
- OAuth support for querying the Github API without a rate limit.
- Automatic PR opening with the `--push` flag.


parsing/parsing_test.go
package parsing

import "testing"

func TestParseAction(t *testing.T) {
	action := parseAction("uses: actions/checkout@v2")
	expected := Action{"actions", "checkout", "2"}
	if action != expected {
		t.Errorf("parseAction: expected %s, got %s", expected, action)
	}
}

func TestRaw(t *testing.T) {
	raw := Action{"actions", "checkout", "2"}.Raw()
	expected := "actions/checkout@v2"
	if raw != expected {
		t.Errorf("Raw: expected %s, got %s", expected, raw)
	}
}

func TestActions(t *testing.T) {
	actions := Actions("uses: actions/checkout@v2\n   uses: actions/cache@v1")
	expected := []Action{
		{"actions", "checkout", "2"},
		{"actions", "cache", "1"}}
	for i, v := range actions {
		if v != expected[i] {
			t.Errorf("Raw: expected %s, got %s", expected[i], v)
		}
	}
}


gitutils/gitutils_test.go
package gitutils

import "testing"

func TestVStrip(t *testing.T) {
	version := versionFormat("v1.2.3")
	if version != "1.2.3" {
		t.Errorf("versionFormat(v1.2.3) ?= 1.2.3, got %s", version)
	}
}


releases/releases_test.go
package releases

import "testing"

func TestVStrip(t *testing.T) {
	version := versionFormat("v1.2.3")
	if version != "1.2.3" {
		t.Errorf("versionFormat(v1.2.3) ?= 1.2.3, got %s", version)
	}
}


config/config.go
package config

import (
	"bufio"
	"context"
	"io/ioutil"
	"os"
	"sync"

	"github.com/fosskers/active/utils"
	"github.com/google/go-github/v31/github"
	"golang.org/x/oauth2"
	"gopkg.in/yaml.v2"
)

// Settings read from a config file.
type Config struct {
	Projects []string `yaml:"projects"`
	Git      Git      `yaml:"git"`
}

type Git struct {
	Name  string `yaml:"name"`
	Email string `yaml:"email"`
	User  string `yaml:"user"`
	Token string `yaml:"token"`
}

// During the lookup of the latest version of an `Action`, we don't want to call
// the Github API more than once per Action. The `seen` map keeps a record of
// lookup attempts.
type Witness struct {
	Seen map[string]bool
	Mut  sync.Mutex
}

// For Actions that actually had a valid 'latest' release, we store the version
// thereof. This is separate from `Witness`, since all _attempted_ lookups might
// not have had an actual result. Keeping them separate also allows for slightly
// less locking.
type Lookups struct {
	Vers map[string]string
	Mut  sync.Mutex
}

// If changes were detected for a given workflow file, we want to prompt the
// user for confirmation before applying them. The update detection process is
// concurrent however, and there would be trouble if multiple prompts appeared
// at the same time.
type Terminal struct {
	Scan *bufio.Scanner
	Mut  sync.Mutex
}

// Our global runtime environment. Passing these as a group simplifies a number
// of function calls. Not every function that receives `Env` will need every
// value, but in practice this isn't a problem.
type Env struct {
	C    *github.Client
	W    *Witness
	L    *Lookups
	T    *Terminal
	Conf *Config
}

// Doesn't mind if the expected fields are missing from the config file.
// Default values are supplied if they are missing.
// Exits the program if parsing the config file failed.
func ReadConfig(path string) *Config {
	c := Config{}
	file, e0 := ioutil.ReadFile(path)
	if e0 == nil {
		e1 := yaml.Unmarshal(file, &c)
		utils.ExitIfErr(e1)
	}
	return &c
}

func GithubClient(config *Config, token *string) *github.Client {
	if *token == "" && config.Git.Token == "" {
		return github.NewClient(nil)
	} else {
		tok := *token
		if tok == "" {
			tok = config.Git.Token
		}
		ctx := context.Background()
		ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: tok})
		tc := oauth2.NewClient(ctx, ts)
		return github.NewClient(tc)
	}
}

// Everything necessary for coordinated concurrency and Github lookups.
func RuntimeEnv(conf *Config, client *github.Client) *Env {
	witness := Witness{Seen: make(map[string]bool)}
	lookups := Lookups{Vers: make(map[string]string)}
	terminal := Terminal{Scan: bufio.NewScanner(os.Stdin)}
	env := Env{client, &witness, &lookups, &terminal, conf}
	return &env
}


parsing/parsing.go
package parsing

import (
	"strings"
)

type Action struct {
	Owner   string
	Name    string
	Version string
}

// The full `owner/repo@vN` format.
func (a Action) Raw() string {
	return a.Repo() + "@v" + a.Version
}

// The `owner/repo` format.
func (a Action) Repo() string {
	return a.Owner + "/" + a.Name
}

// Given the contents of a workflow YAML file, find all uses of a Github Action.
func Actions(file string) []Action {
	lines := strings.Split(file, "\n")
	actions := make([]Action, 0, len(lines))
	for _, line := range lines {
		l := strings.TrimSpace(line)
		if strings.HasPrefix(l, "uses:") {
			action := parseAction(l)
			actions = append(actions, action)
		}
	}
	return actions
}

// Form an `Action`, given a line like:
//
//      uses: actions/checkout@v2
func parseAction(line string) Action {
	dropped := line[6:]
	owner := strings.SplitN(dropped, "/", 2)
	name := strings.SplitN(owner[1], "@", 2)
	return Action{Owner: owner[0], Name: name[0], Version: name[1][1:]}
}


gitutils/gitutils.go
package gitutils

import (
	"context"
	"fmt"
	"path/filepath"
	"strings"
	"time"

	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/config"
	"github.com/go-git/go-git/v5/plumbing"
	"github.com/go-git/go-git/v5/plumbing/object"
	"github.com/go-git/go-git/v5/plumbing/transport/http"
	"github.com/google/go-github/v31/github"
)

// Given an activated client and a Github project, look up the version of its
// most recent release.
func Recent(client *github.Client, owner, repo string) (string, error) {
	rel, _, err := client.Repositories.GetLatestRelease(context.Background(), owner, repo)
	if err != nil {
		return "", err
	}
	return versionFormat(rel.GetTagName()), nil
}

// Strip the `v` from the beginning of the tag name.
func versionFormat(version string) string {
	return version[1:]
}

// Switch to a given branch.
func Checkout(r *git.Repository, branch string) error {
	w, e0 := r.Worktree()
	if e0 != nil {
		return e0
	}

	e1 := w.Checkout(&git.CheckoutOptions{
		Branch: plumbing.ReferenceName("refs/heads/" + branch),
	})
	if e1 != nil {
		return e1
	}

	return nil
}

// Create a new branch and switch to it, based off the current branch.
func CheckoutCreate(r *git.Repository, branch string) error {
	ref, e0 := r.Head()
	if e0 != nil {
		return e0
	}

	w, e1 := r.Worktree()
	if e1 != nil {
		return e1
	}

	e2 := w.Checkout(&git.CheckoutOptions{
		Hash:   ref.Hash(),
		Branch: plumbing.ReferenceName("refs/heads/" + branch),
		Create: true,
	})
	if e2 != nil {
		return e2
	}

	return nil
}

// Commit the changes in some given filepaths.
func Commit(r *git.Repository, name string, email string, files []string) error {
	w, e0 := r.Worktree()
	if e0 != nil {
		return e0
	}
	for _, file := range files {
		_, e1 := w.Add(file)
		if e1 != nil {
			return e1
		}
	}
	_, e2 := w.Commit("[active] Updating Github Actions", &git.CommitOptions{
		Author: &object.Signature{Name: name, Email: email, When: time.Now()},
	})
	if e2 != nil {
		return e2
	}

	return nil
}

// Push the given branch.
func Push(r *git.Repository, remote string, branch string, user string, token string) error {
	src := filepath.Join("refs/heads/", branch)
	spec := config.RefSpec(src + ":" + src)
	return r.Push(&git.PushOptions{
		RemoteName: remote,
		RefSpecs:   []config.RefSpec{spec},
		Auth:       &http.BasicAuth{Username: user, Password: token},
	})
}

// Pull the `master` branch.
func PullMaster(w *git.Worktree, remote string, user string, token string) error {
	return w.Pull(&git.PullOptions{
		RemoteName:    remote,
		ReferenceName: plumbing.Master,
		SingleBranch:  true,
		Auth:          &http.BasicAuth{Username: user, Password: token},
	})
}

// Open a pull request, and return its number.
func PullRequest(c *github.Client, owner string, repo string, branch string) (int, error) {
	new := &github.NewPullRequest{
		Title:               github.String("Github CI Action Updates"),
		Head:                github.String(branch),
		Base:                github.String("master"),
		Body:                github.String("This PR was opened automatically by the `active` tool."),
		MaintainerCanModify: github.Bool(true),
	}
	pr, _, e0 := c.PullRequests.Create(context.Background(), owner, repo, new)
	if e0 != nil {
		return 0, e0
	}
	return *pr.Number, nil
}

// Pick a suitable "remote" to push to later, defaulting to one that was created
// previously by this tool.
func ChooseRemote(rs []*git.Remote) *git.Remote {
	for _, r := range rs {
		if r.Config().Name == "active" {
			return r
		}
	}
	for _, r := range rs {
		if r.Config().Name == "origin" {
			return r
		}
	}
	if len(rs) > 0 {
		return rs[0]
	}
	return nil
}

// Fetch or create an HTTP-based remote that we can used to push via the given
// Github API token. Yields the name of the remote, as well as the owner of the
// repo.
func PushableRemote(repo *git.Repository) (string, string, error) {
	rs, e0 := repo.Remotes()
	if e0 != nil {
		return "", "", e0
	}

	chosen := ChooseRemote(rs)
	if chosen == nil {
		return "", "", fmt.Errorf("No remotes found.")
	}

	rc := chosen.Config()
	if rc == nil {
		return "", "", fmt.Errorf("Couldn't fetch RemoteConfig.")
	}

	if len(rc.URLs) == 0 {
		return "", "", fmt.Errorf("Given remote had no URLs!")
	}

	// We don't need to create a new remote; the one given uses HTTPS already.
	if rc.URLs[0][0:5] == "https" {
		owner := strings.Split(rc.URLs[0][8:], "/")[1]
		return rc.Name, owner, nil
	}

	base := "https://" + strings.ReplaceAll(rc.URLs[0][4:], ":", "/")
	owner := strings.Split(base[8:], "/")[1]

	new := config.RemoteConfig{
		Name: "active",
		URLs: []string{base},
	}

	_, e1 := repo.CreateRemote(&new)
	if e1 != nil {
		return "", "", e1
	}

	return "active", owner, nil
}


active.go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/fatih/color"
	"github.com/fosskers/active/config"
	"github.com/fosskers/active/gitutils"
	"github.com/fosskers/active/parsing"
	"github.com/fosskers/active/utils"
	"github.com/go-git/go-git/v5"
	"github.com/google/go-github/v31/github"
)

// Paths.
var home, _ = os.UserHomeDir()
var confPath string = filepath.Join(home, ".config/active.yaml")

// Command-line flags.
var localF *bool = flag.Bool("local", false, "Check the local repo you're currently in.")
var tokenF *string = flag.String("token", "", "(optional) Github API OAuth Token.")
var autoF *bool = flag.Bool("y", false, "Automatically apply changes.")
var configPathF *string = flag.String("config", confPath, "Path to config file.")
var pushF *bool = flag.Bool("push", false, "Automatically make commits and open a PR on Github.")
var nocolourF *bool = flag.Bool("nocolor", false, "Disable coloured output.")

// Coloured output.
var cyan = color.New(color.FgCyan).SprintFunc()
var yellow = color.New(color.FgYellow).SprintFunc()
var green = color.New(color.FgGreen).SprintFunc()

type Project struct {
	name      string
	owner     string
	remote    string
	workflows []*Workflow
	repo      *git.Repository
	accepted  []string // Mutable field.
	branch    string
}

// All data pertaining to a fully read and parsed Workflow file.
type Workflow struct {
	path    string // Full filepath to the workflow file.
	yaml    string
	actions []parsing.Action
}

func main() {
	flag.Parse()                         // Collect command-line options.
	c := config.ReadConfig(*configPathF) // Read the config file.

	if *nocolourF {
		color.NoColor = true
	}

	if *pushF && *tokenF == "" && c.Git.Token == "" {
		utils.PrintExit("A real token must be given when using '--push'.")
	}

	client := config.GithubClient(c, tokenF) // Github communication.
	env := config.RuntimeEnv(c, client)      // Runtime environment.
	projects := allProjects(c)

	// Report discovered files.
	longest := 0
	for _, proj := range projects {
		projlen := len(proj.name)
		if projlen > longest {
			longest = projlen
		}
	}
	fmt.Println("Checking the following files:")
	for _, proj := range projects {
		for _, w := range proj.workflows {
			spaces := strings.Repeat(" ", longest-len(proj.name))
			fmt.Printf("  --> %s: %s%s\n", cyan(proj.name), spaces, filepath.Base(w.path))
		}
	}

	// Register parsed Actions (calls the Github API).
	var wg sync.WaitGroup
	for _, proj := range projects {
		for _, wf := range proj.workflows {
			wg.Add(1)
			go func(w *Workflow) {
				register(env, w.actions)
				wg.Done()
			}(wf)
		}
	}
	wg.Wait()

	// Perform updates concurrently.
	for _, proj := range projects {
		wg.Add(1)
		go func(p *Project) {
			applyUpdates(env, p)
			wg.Done()
		}(proj)
	}
	wg.Wait()

	// Commit and push updates to Github.
	if *pushF {
		for _, proj := range projects {
			if len(proj.accepted) > 0 {
				wg.Add(1)
				go func(p *Project) {
					defer wg.Done()
					defer gitutils.Checkout(p.repo, "master")
					pr, e := commitAndPush(client, c, p)
					if e != nil {
						fmt.Println(e)
						return
					}
					fmt.Printf("Successfully opened a PR for %s! (#%d)\n", cyan(p.name), pr)
				}(proj)
			}
		}
		wg.Wait()
	}

	fmt.Println("Done.")
}

// Will exit the program if there are no projects to check, or if a specified
// project has no workflow files.
func allProjects(c *config.Config) []*Project {
	if *localF {
		p, e0 := project(c, ".")
		utils.ExitIfErr(e0) // Fail hard if the only project we're checking is invalid.
		return []*Project{p}
	}

	if len(c.Projects) == 0 {
		utils.PrintExit("No projects to check. Try '--local' or setting your config file.")
	}

	var wg sync.WaitGroup
	var mut sync.Mutex
	ps := make([]*Project, 0)
	for _, proj := range c.Projects {
		wg.Add(1)
		go func(p string) {
			defer wg.Done()
			proj, e0 := project(c, p)
			if e0 != nil {
				fmt.Println(e0)
				return
			}
			mut.Lock()
			ps = append(ps, proj)
			mut.Unlock()
		}(proj)
	}
	wg.Wait()
	return ps
}

// Given a local path to a Git repository, read everything from the filesystem
// that's necessary for further processing.
//
// Exits the program if even one file fails to be read, or if there weren't any
// to be read for the given project.
func project(c *config.Config, path string) (*Project, error) {
	name := filepath.Base(path)

	var repo *git.Repository
	owner := ""
	remote := ""
	branch := ""
	if *pushF {
		r, e0 := git.PlainOpen(path)
		if e0 != nil {
			return nil, e0
		}
		repo = r

		rem, own, e1 := gitutils.PushableRemote(r)
		if e1 != nil {
			return nil, e1
		}
		remote = rem
		owner = own

		br, e2 := switchBranches(c, r, remote, name)
		if e2 != nil {
			return nil, e2
		}
		branch = br
	}

	// Read and parse all Workflow files.
	wps, e1 := workflows(path)
	utils.ExitIfErr(e1)
	if len(wps) == 0 {
		utils.PrintExit("No workflow files detected for " + name)
	}
	ws := make([]*Workflow, 0)
	for _, wp := range wps {
		yaml := readWorkflow(wp)
		actions := parsing.Actions(yaml)
		workflow := Workflow{wp, yaml, actions}
		ws = append(ws, &workflow)
	}

	return &Project{
		name:      name,
		owner:     owner,
		remote:    remote,
		workflows: ws,
		repo:      repo,
		accepted:  make([]string, 0),
		branch:    branch,
	}, nil
}

// Given a local path to a code repository, find the paths of all its Github
// workflow configuration files.
func workflows(project string) ([]string, error) {
	workflowDir := filepath.Join(project, ".github/workflows")
	items, err := ioutil.ReadDir(workflowDir)
	if err != nil {
		return nil, err
	}
	fullPaths := make([]string, 0, len(items))
	for _, file := range items {
		if !file.IsDir() {
			full := filepath.Join(workflowDir, file.Name())
			fullPaths = append(fullPaths, full)
		}
	}
	return fullPaths, nil
}

// Detect and apply updates.
func applyUpdates(env *config.Env, project *Project) {
	// ASSUMPTION: `env.L.Vers` has been fully written to, and will only be read
	// from here on.
	ls := env.L.Vers

	// Apply updates, if the user wants them.
	for _, wf := range project.workflows {
		newAs := newActionVers(ls, wf.actions)
		yamlNew := update(newAs, wf.yaml)

		// Only proceed if there were actually changes to consider.
		if wf.yaml != yamlNew {
			env.T.Mut.Lock()
			resp := prompt(env, project.name, wf, newAs)

			if resp {
				ioutil.WriteFile(wf.path, []byte(yamlNew), 0644)
				fmt.Println("Updated.")

				// Mutability to communicate back to `main` that the user
				// accepted these changes.
				path := filepath.Join(".github/workflows", filepath.Base(wf.path))
				project.accepted = append(project.accepted, path)
			} else {
				fmt.Println("Skipping...")
			}
			env.T.Mut.Unlock()
		}
	}
}

// Switch git branches, if we haven't already. go-git does not
// support stashing, so if the working tree isn't clean, we have
// to skip this Project entirely. This also pulls the latest master from
// the remote.
func switchBranches(c *config.Config, r *git.Repository, remote string, pname string) (string, error) {
	wt, e9 := r.Worktree()
	if e9 != nil {
		return "", e9
	}

	status, e8 := wt.Status()
	if e8 != nil {
		return "", e8
	}

	if !status.IsClean() {
		return "", fmt.Errorf("The working tree of %s is not clean.", cyan(pname))
	}
	e0 := gitutils.Checkout(r, "master")
	if e0 != nil {
		return "", fmt.Errorf("Unable to switch branches for %s: %s", cyan(pname), e0)
	}
	e2 := gitutils.PullMaster(wt, remote, c.Git.User, c.Git.Token)
	if e2 != nil && e2 != git.NoErrAlreadyUpToDate {
		return "", fmt.Errorf("Could not pull master for %s: %s", cyan(pname), e2)
	}
	branch := "active/" + time.Now().Format("2006-01-02-15-04-05")
	e1 := gitutils.CheckoutCreate(r, branch)
	if e1 != nil {
		return "", fmt.Errorf("Unable to create a new branch for %s: %s", cyan(pname), e1)
	}
	return branch, nil
}

// Read the workflow file, if we can. Exit otherwise, since the user
// probably wasn't expecting that their file was unreadable.
func readWorkflow(path string) string {
	yamlRaw, err := ioutil.ReadFile(path)
	utils.ExitIfErr(err)
	return string(yamlRaw)
}

// Given some Actions, call the Github API and check for their latest versions.
func register(env *config.Env, actions []parsing.Action) {
	var wg sync.WaitGroup
	for _, action := range actions {
		wg.Add(1)
		go func(action parsing.Action) {
			versionLookup(env, action)
			wg.Done()
		}(action)
	}
	wg.Wait()
}

// Concurrently query the Github API for Action versions.
func versionLookup(env *config.Env, a parsing.Action) {
	// Have we looked up this Action already?
	env.W.Mut.Lock()
	repo := a.Repo()
	if seen := env.W.Seen[repo]; seen {
		env.W.Mut.Unlock()
		return
	}
	env.W.Seen[repo] = true
	env.W.Mut.Unlock()

	// Version lookup and recording.
	version, err := gitutils.Recent(env.C, a.Owner, a.Name)
	if err != nil {
		return
	}
	env.L.Mut.Lock()
	env.L.Vers[repo] = version
	env.L.Mut.Unlock()
}

// For some Actions, what new version should they be assigned to?
func newActionVers(ls map[string]string, actions []parsing.Action) map[parsing.Action]string {
	news := make(map[parsing.Action]string)
	for _, action := range actions {
		if v := ls[action.Repo()]; v != "" && action.Version != v {
			news[action] = v
		}
	}
	return news
}

// Given the Actions detected in some workflow file, try to replace them with
// the newest versions available from Github.
func update(actions map[parsing.Action]string, yaml string) string {
	yamlNew := yaml
	for action, v := range actions {
		old := "uses: " + action.Raw()
		new := "uses: " + action.Repo() + "@v" + v
		yamlNew = strings.ReplaceAll(yamlNew, old, new)
	}
	return yamlNew
}

// We detected some changes to a workflow file, so we inform the user and ask
// whether we should write the changes to disk.
func prompt(env *config.Env, projName string, workflow *Workflow, newAs map[parsing.Action]string) bool {
	longestName := 0
	longestVer := 0
	for action := range newAs {
		if repo := action.Repo(); len(repo) > longestName {
			longestName = len(repo)
		}
		if len(action.Version) > longestVer {
			longestVer = len(action.Version)
		}
	}
	fmt.Printf("\nUpdates available for %s: %s:\n", cyan(projName), filepath.Base(workflow.path))
	for action, v := range newAs {
		repo := action.Repo()
		nameDiff := longestName - len(repo)
		verDiff := longestVer - len(action.Version)
		spaces := strings.Repeat(" ", nameDiff+verDiff+1)
		patt := "  %s" + spaces + "%s --> %s\n"
		fmt.Printf(patt, repo, yellow(action.Version), green(v))
	}

	resp := "NO"
	if !*autoF {
		fmt.Printf("Would you like to apply them? [Y/n] ")
		env.T.Scan.Scan()
		resp = env.T.Scan.Text()
	}
	return *autoF || resp == "Y" || resp == "y" || resp == ""
}

// Attempt to commit the changes, push the branch, and open a new PR.
// The yielded int is the number of the new PR, if opened.
func commitAndPush(client *github.Client, c *config.Config, p *Project) (int, error) {
	e0 := gitutils.Commit(p.repo, c.Git.Name, c.Git.Email, p.accepted)
	if e0 != nil {
		return 0, fmt.Errorf("Couldn't commit %s: %s\n", cyan(p.name), e0)
	}
	e1 := gitutils.Push(p.repo, p.remote, p.branch, c.Git.User, c.Git.Token)
	if e1 != nil {
		return 0, fmt.Errorf("Unable to push %s to Github: %s\n", cyan(p.name), e1)
	}
	pr, e2 := gitutils.PullRequest(client, p.owner, p.name, p.branch)
	if e2 != nil {
		return 0, fmt.Errorf("Opening a PR for %s failed: %s\n", cyan(p.name), e2)
	}
	return pr, nil
}
